zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1028,14 +1019,21 @@ void primeConnection() throws IOException {
                             } else if (childWatchesIter.hasNext()) {
                                 watch = childWatchesIter.next();
                                 childWatchesBatch.add(watch);
+                            }  else if (persistentWatchesIter.hasNext()) {
+                                watch = persistentWatchesIter.next();
+                                persistentWatchesBatch.add(watch);
+                            } else if (persistentRecursiveWatchesIter.hasNext()) {
+                                watch = persistentRecursiveWatchesIter.next();
+                                persistentRecursiveWatchesBatch.add(watch);
                             } else {
                                 break;
                             }
                             batchLength += watch.length();
                         }
 
-                        SetWatches sw = new SetWatches(setWatchesLastZxid, dataWatchesBatch, existWatchesBatch, childWatchesBatch);
-                        RequestHeader header = new RequestHeader(-8, OpCode.setWatches);
+                        SetWatches2 sw = new SetWatches2(setWatchesLastZxid, dataWatchesBatch, existWatchesBatch,
```

Comment:
How to deal with old servers? Should we have a configuration flag?

Reply:
Hmm - good point. I'll add that. Are there other examples of this?

Reply:
It's only an issue for the client, so maybe add a method to `ZooKeeper()`, something like `oldServerMode()` (name TBD) that causes SetWatches to be used instead of SetWatches2.

Reply:
I think it is better to use the new message type when it is required, that is when you have at least one persistent watch

Reply:
That's a good solution. I've just pushed this change.